iothub/device/src/DeviceClient.cs

```diff
@@ -447,9 +449,28 @@ public void SetRetryPolicy(IRetryPolicy retryPolicy)
         /// <param name="cancellationToken">A cancellation token to cancel the operation.</param>
         /// <exception cref="OperationCanceledException">Thrown when the operation has been canceled.</exception>
         /// <returns>The task to await</returns>
+        [Obsolete("This API has been split into three APIs: GetFileUploadSasUri, uploading to blob directly using the Azure Storage SDK, and CompleteFileUploadAsync")]
         public Task UploadToBlobAsync(string blobName, Stream source, CancellationToken cancellationToken) =>
             InternalClient.UploadToBlobAsync(blobName, source, cancellationToken);
 
+        /// <summary>
+        /// Get a file upload SAS URI which the Azure Storage SDK can use to upload a file to blob for this device. See <see href="https://docs.microsoft.com/en-us/azure/iot-hub/iot-hub-devguide-file-upload#initialize-a-file-upload">this documentation for more details</see>
+        /// </summary>
+        /// <param name="request">The request details for getting the SAS URI, including the destination blob name.</param>
+        /// <param name="cancellationToken">The cancellation token.</param>
+        /// <returns>The file upload details to be used with the Azure Storage SDK in order to upload a file from this device.</returns>
+        public Task<FileUploadSasUriResponse> GetFileUploadSasUri(FileUploadSasUriRequest request, CancellationToken cancellationToken = default) =>
+            InternalClient.GetFileUploadSasUriAsync(request, cancellationToken);
+
+        /// <summary>
+        /// Notify IoT Hub that a device's file upload has finished. See <see href="https://docs.microsoft.com/en-us/azure/iot-hub/iot-hub-devguide-file-upload#notify-iot-hub-of-a-completed-file-upload">this documentation for more details</see>
+        /// </summary>
+        /// <param name="notification">The notification details, including if the file upload succeeded.</param>
+        /// <param name="cancellationToken">The cancellation token.</param>
+        /// <returns>The task to await.</returns>
+        public Task CompleteFileUploadAsync(FileUploadCompletionNotification notification, CancellationToken cancellationToken = default) =>
```

Comment:
Can a customer reject this? (to be inline with our other API's)

Reply:
No, because it is not like those other APIs. The device client is the sender, not the receiver

Reply:
The proper method name might be "NotifyHubThatFileUploadHasCompleted" but that is a little wordy

Reply:
But as a user I can decide to not upload after receiving sas uri. How do I notify that?

Reply:
You can decide to not upload, and there is nothing wrong with that. You can even notify iothub that an upload occurred even when it didn't, but that would just be user error.